src/main/java/com/puppycrawl/tools/checkstyle/meta/MetadataGeneratorUtil.java

```diff
@@ -105,4 +157,192 @@ private static List<File> getTargetFiles(String path, String... moduleFolders)
+                moduleType, propertiesDetails, new ArrayList<>(messageKeys));
+    }
+
+    /**
+     * Get module type(check/filter/filefilter) based on module name.
+     *
+     * @param moduleName module name.
+     * @return module type.
+     */
+    private static ModuleType getModuleType(String moduleName) {
+        final ModuleType result;
+        if (moduleName.endsWith("FileFilter")) {
+            result = ModuleType.FILEFILTER;
+        }
+        else if (moduleName.endsWith("Filter")) {
+            result = ModuleType.FILTER;
+        }
+        else {
+            result = ModuleType.CHECK;
+        }
+        return result;
+    }
+
+    /**
+     * Get property details for the given property - name, description, type, default value.
+     *
+     * @param properties properties of the module.
+     * @param javadocs javadocs of the module.
+     * @param className the class name of the module.
+     * @param instance the instance of the module.
+     * @return property details.
+     * @throws MacroExecutionException if an error occurs.
+     */
+    private static List<ModulePropertyDetails> getPropertiesDetails(
+            Set<String> properties, Map<String, DetailNode> javadocs,
+            String className, Object instance)
+            throws MacroExecutionException {
+        final List<ModulePropertyDetails> result = new ArrayList<>(properties.size());
+        for (String property : properties) {
+            final String description = getPropertyDescription(property,
+                    javadocs.get(property));
+            final Field field = SiteUtil.getField(instance.getClass(), property);
+            final String type =
+                    SiteUtil.getFieldClass(field, property, className, instance).getCanonicalName();
+            final String defaultValue = getPropertyDefaultValue(property, field, instance,
+                    className);
+            final String validationType = getValidationType(property);
+
+            result.add(new ModulePropertyDetails(property, type, defaultValue,
+                    validationType, description));
+        }
+        return result;
+    }
+
+    /**
+     * Get validation type for the given property.
+     *
+     * @param property property name.
+     * @return validation type.
+     */
+    private static String getValidationType(String property) {
+        final String validationType;
+        if (SiteUtil.TOKENS.equals(property) || SiteUtil.JAVADOC_TOKENS.equals(property)) {
+            validationType = "tokenSet";
+        }
+        else {
+            validationType = null;
+        }
+        return validationType;
+    }
+
+    /**
+     * Get property description from property javadoc.
+     *
+     * @param property property name.
+     * @param propertyJavadoc property javadoc.
+     * @return property description.
+     */
+    private static String getPropertyDescription(String property, DetailNode propertyJavadoc) {
```

Comment:
I am not sure why this method is specific for metadata.
I think there should single method that does it for xdoc and meta and meta generation just reuse it 

If agree please create separate issue to improve in separate PR

Reply:
If you agree add it to https://github.com/checkstyle/checkstyle/issues/13699

Reply:
I think there should be separate methods for xdoc and metadata. Or at least single method with different output depending on who/what is calling it.

Metadata gets text as-is and doesn't care about any processing.

Xdoc requires to 
- convert `{@code something}` to `<code>something</code>`
- Convert links to relative ones-  `https://checkstyle.org/checks/block/xxx.html` -> `../block/xx.html`, for example

Reply:
But still data creation should be single method and xdoc just do only such required formatting. If you agree just put this to separate issue that we have for design improvement as item.